lib/hocon/impl/resolve_context.rb

```diff
@@ -27,6 +29,32 @@ def self.new_cycle_markers
     Set.new
   end
 
+  def add_cycle_marker(value)
+    if Hocon::Impl::ConfigImpl.trace_substitution_enabled
+      Hocon::Impl::ConfigImpl.trace("++ Cycle marker " + value + "@" + value.hash,
+                       depth)
+    end
+    if @cycle_markers.include?(value)
+      raise ConfigBugOrBrokenError.new("Added cycle marker twice " + value)
+    end
+    copy = self.class.new_cycle_markers
+    copy.merge(@cycle_markers)
+    copy.add(value)
+    self.class.new(@memos, @options, @restrict_to_child, @resolve_stack, copy)
+  end
+
+  def remove_cycle_marker(value)
+    if Hocon::Impl::ConfigImpl.trace_substitution_enabled
+      Hocon::Impl::ConfigImpl.trace("++ Cycle marker " + value + "@" + value.hash,
```

Comment:
Probably not important, but I noticed that the java code says `-- Cycle marker`, with hyphens